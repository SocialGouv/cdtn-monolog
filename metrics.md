## Metrics:
---
Different metrics are computed to collect insights on the website usage and generate report in the following format:

```r
metricsAnalysis = {
        "nbVisitsAnalyzed" : number of user sessions analyzed on the given oeriod,
        "longVisitsNb": Number of Long visits defined as:
         - either user visits more than 2 "visit_content" pages
         - or uses one of the "Search", "Themes", "selectRelated" features in their session.
        "longVisitsRatio" : LongVisitsNb/nbVisitsAnalyzed, 
        "shortVisitsNb" : should be equal to nbVisitsAnalyzed - LongVisitsNb, 

        "SelectRelatedCount" : nb of users whose session includes the selectRelated feature,
        "visitorSelectedRelatedRatio" : SelectRelatedCount/nbVisitsAnalyzed.
        "visitorWasRedirected": nb of users whose session includes the selectRelated feature AND came on the website via a search engine.
        "reportId": id of the report based on Date,
    }
```