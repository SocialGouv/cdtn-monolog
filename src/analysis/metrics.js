import * as datasetUtil from "../dataset";
import * as util from "../util";

const reportType = "metrics";

const removeAnchor = (url) => {
  return url.split("#")[0];
};

const removeQuery = (url) => {
  return url.split("?q=")[0];
};

const add = (a, b) => a + b;
const avg = (array) => {
  const sum = array.reduce((previous, current) => (current += previous));
  return sum / array.length;
};

const valueCounts = (arr) => {
  var counts = {};

  for (var i = 0; i < arr.length; i++) {
    var num = arr[i];
    counts[num] = counts[num] ? counts[num] + 1 : 1;
  }
  return counts;
};

const typeCounts = (visit) => {
  return visit
    .groupBy((row) => row.type)
    .select((group) => {
      return {
        count: group.count(),
        referrerTypeName: group.first().referrerTypeName,
        type: group.first().type,
      };
    });
};

const isExploVisit = (x) => {
  // if more than 2 contents visited --> explo
  // else: if search or themes used --> explo
  if (x.type == "visit_content") {
    return x.count > 2 ? true : false;
  } else {
    const exploTypes = Array("search", "themes", "selectRelated");
    if (exploTypes.includes(x.type)) {
      return true;
    } else {
      return false;
    }
  }
};

const isRedirected = (x) => {
  // if a user comes from a search engine and select related return true
  if (x.referrerTypeName == "Search Engines" && x.type == "selectRelated") {
    return true;
  } else {
    return false;
  }
};
const hasSelectedRelated = (x) => {
  return x.type == "selectRelated" ? true : false;
};
const countSelectRelated = (x) => {
  return x.type == "selectRelated" ? x.count : 0;
};
const getSelectRelated = (relatedCount) => {
  // takes an array with true if eventType selectRelated else false
  // returns two metric per session:
  // visitorSelectedRelated true if the user clicked at least one on selectRelated
  // SelectRelatedCount: the nb of times the user selected related contents

  return {
    selectRelatedCount: relatedCount
      .map((x) => countSelectRelated(x))
      .reduce(add),
    visitorSelectedRelated: relatedCount.some(hasSelectedRelated),
    visitorWasRedirected: relatedCount.some(isRedirected),
  };
};
const getUserType = (visitCount) => {
  // based on visit counts [{'type':visit_content, 'count': 2}]
  //determine whether its long or short visits removing nas
  return visitCount.some(isExploVisit);
};
const clean = (visits) => {
  const seriesOfVisits = visits.toArray();
  // remove sections and remove queries from urls
  const cleanSeriesOfVisits = seriesOfVisits.map((x) =>
    x.transformSeries({
      url: (u) => util.urlToPath(removeQuery(removeAnchor(u))),
    })
  );
  // deduplicates urls within user session (reloads)
  const uniqueSeriesOfVisits = cleanSeriesOfVisits.map((x) =>
    x.distinct((x) => {
      x.url, x.type;
    })
  );
  return uniqueSeriesOfVisits.map((visit) => typeCounts(visit).toArray());
};
const analyse = (dataset, reportId) => {
  const visits = datasetUtil.getVisits(dataset); //series of dataframes

  // CLEANING

  // count visits by type of events (give a list of dict [{'type':visit_content, 'count': 2}, {}])
  const visitsTypesArray = clean(visits);

  // ANALYSIS
  const nbVisitsAnalyzed = visitsTypesArray.length;

  // SelectRelated
  const selectRelatedStats = visitsTypesArray.map((x) => getSelectRelated(x));

  const visitorSelectedRelated = selectRelatedStats.map(
    (x) => x["visitorSelectedRelated"]
  );
  const SelectRelatedCount = selectRelatedStats.map(
    (x) => x["selectRelatedCount"]
  );
  const visitorWasRedirected = selectRelatedStats.map(
    (x) => x["visitorWasRedirected"]
  );

  //visitorTypes
  const isLongVisitArray = visitsTypesArray.map((x) => getUserType(x)); // return false true array if user is long
  const LongVisitRatio = avg(isLongVisitArray); // simple average of the visitcount array
  const visitsTypesCount = valueCounts(isLongVisitArray);

  const metricsAnalysis = {
    SelectRelatedCount: SelectRelatedCount.reduce(add),
    longVisitsNb: visitsTypesCount[false],
    longVisitsRatio: LongVisitRatio,
    nbVisitsAnalyzed: nbVisitsAnalyzed,
    reportId: reportId,
    reportType,
    shortVisitsNb: visitsTypesCount[true],
    visitorSelectedRelatedRatio: avg(visitorSelectedRelated),
    visitorWasRedirected: visitorWasRedirected.reduce(add),
  };
  return Array(metricsAnalysis);
};

export {
  analyse,
  reportType,
  removeAnchor,
  removeQuery,
  add,
  avg,
  valueCounts,
  typeCounts,
  isExploVisit,
  isRedirected,
  hasSelectedRelated,
  countSelectRelated,
  getSelectRelated,
  getUserType,
  clean,
};
