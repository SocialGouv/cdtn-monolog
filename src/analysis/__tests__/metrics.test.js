import * as dataForge from "data-forge";

import * as Metrics from "../metrics";

// test minimal utils

describe("Cleaning", () => {
  it("should remove anchors in url", () => {
    const url = "https://whatever.com/content#section";
    const base_url = "https://whatever.com/content";
    const url_no_anchor = Metrics.removeAnchor(url);
    expect(url_no_anchor).toBe(base_url);
  });
  it("should remove queries in url", () => {
    const url = "https://whatever.com/content?q=somequery'('é('é@&§§('";
    const base_url = "https://whatever.com/content";
    const url_no_query = Metrics.removeQuery(url);
    expect(url_no_query).toBe(base_url);
  });
  it("should remove both queries and anchors in url", () => {
    const url = "https://whatever.com/content#anchor?q=somequery";
    const base_url = "https://whatever.com/content";
    const url_cleaned = Metrics.removeQuery(Metrics.removeAnchor(url));
    expect(url_cleaned).toBe(base_url);
  });
});

describe("Cleaning pipe", () => {
  it("shodl clean the user logs correctly", () => {
    const dirtyVisits1 = new dataForge.DataFrame([
      {
        type: "visit_content",
        referrerTypeName: "Search Engines",
        url: "https://whatever.com/content#anchor?q=somequery",
      },
      {
        type: "visit_content",
        referrerTypeName: "Search Engines",
        url: "https://whatever.com/duplicate?q=somequery",
      },
      {
        type: "visit_content",
        referrerTypeName: "Search Engines",
        url: "https://whatever.com/notduplicate?q=somequery",
      },
      {
        type: "outils",
        referrerTypeName: "Search Engines",
        url: "https://whatever.com/duplicate#anothersection?q=anotherquery",
      },
    ]);
    const dirtyVisits2 = new dataForge.DataFrame([
      {
        type: "visit_content",
        referrerTypeName: "Search Engines",
        url: "another-user.com/content",
      },
    ]);
    const dirtyVisits = new dataForge.Series([dirtyVisits1, dirtyVisits2]);
    const cleanedSeries = Metrics.clean(dirtyVisits);
    //console.log(cleanedSeries);
    expect(cleanedSeries).toMatchSnapshot();
  });
});

describe("Basic arithmetic", () => {
  it("should sum bolean elements of array", () => {
    const arr = Array(true, false, true);
    const sum = arr.reduce(Metrics.add);
    expect(sum).toBe(2);
  });
  it("should sum average elements of array", () => {
    const arr = Array(true, false, true);
    const average = Metrics.avg(arr);
    expect(average).toBe(2 / 3);
  });
  it("should count nb of elements of each value  in array", () => {
    const arr = Array(true, false, true, "whatever");
    const values = Metrics.valueCounts(arr);
    expect(values).toStrictEqual({ false: 1, true: 2, whatever: 1 });
  });
});

describe("Business logic", () => {
  it("it should aggregates user sessions", () => {
    const session = new dataForge.DataFrame([
      { referrerTypeName: null, type: "visit_content" },
      { referrerTypeName: null, type: "visit_content" },
      { referrerTypeName: null, type: "visit_content" },
    ]);
    const sessionsCounts = Metrics.typeCounts(session);
    expect(sessionsCounts.toArray()).toStrictEqual([
      { count: 3, referrerTypeName: null, type: "visit_content" },
    ]);
  });
  it("it should aggregates user having different content types", () => {
    const session = new dataForge.DataFrame([
      { referrerTypeName: null, type: "visit_content" },
      { referrerTypeName: null, type: "visit_content" },
      { referrerTypeName: null, type: "selectRelated" },
      { referrerTypeName: null, type: "selectRelated" },
      { referrerTypeName: null, type: "selectRelated" },
    ]);
    const sessionsCounts = Metrics.typeCounts(session);
    expect(sessionsCounts.toArray()).toStrictEqual([
      { count: 2, referrerTypeName: null, type: "visit_content" },
      { count: 3, referrerTypeName: null, type: "selectRelated" },
    ]);
  });
  it("it should detect exploratory visits", () => {
    const sessionCount = {
      count: 5,
      referrerTypeName: null,
      type: "visit_content",
    };
    const isExplo = Metrics.isExploVisit(sessionCount);
    expect(isExplo).toBe(true);
  });
  it("it should detect exploratory visits search", () => {
    const sessionCount = { count: 1, referrerTypeName: null, type: "search" };
    const isExplo = Metrics.isExploVisit(sessionCount);
    expect(isExplo).toBe(true);
  });
  it("it should detect exploratory visits themes", () => {
    const sessionCount = { count: 1, referrerTypeName: null, type: "themes" };
    const isExplo = Metrics.isExploVisit(sessionCount);
    expect(isExplo).toBe(true);
  });
  it("it should detect exploratory visits selectResult", () => {
    const sessionCount = {
      count: 1,
      referrerTypeName: null,
      type: "selectRelated",
    };
    const isExplo = Metrics.isExploVisit(sessionCount);
    expect(isExplo).toBe(true);
  });
  it("it should detect empty visits as short", () => {
    const sessionCount = {};
    const isExplo = Metrics.isExploVisit(sessionCount);
    expect(isExplo).toBe(false);
  });
  it("it should not detect visits with noise", () => {
    const sessionCount = {
      count: 100,
      referrerTypeName: null,
      type: "whatever",
    };
    const isExplo = Metrics.isExploVisit(sessionCount);
    expect(isExplo).toBe(false);
  });
  it("it should detect redirected user", () => {
    const sessionCount = {
      count: 1,
      referrerTypeName: "Search Engines",
      type: "selectRelated",
    };
    const isRedirect = Metrics.isRedirected(sessionCount);
    expect(isRedirect).toBe(true);
  });
  it("it shouldnot has selected related", () => {
    const sessionCount = {
      count: 1,
      referrerTypeName: "Direct Entry",
      type: "selectRelated",
    };
    const isRedirect = Metrics.hasSelectedRelated(sessionCount);
    expect(isRedirect).toBe(true);
  });
  it("it should count selectRelated", () => {
    const sessionCount = {
      count: 8,
      referrerTypeName: "Direct Entry",
      type: "selectRelated",
    };
    const countrelated = Metrics.countSelectRelated(sessionCount);
    expect(countrelated).toBe(8);
  });
  it("it should get stats of selectRelated", () => {
    const sessionCountArray = [
      { count: 8, referrerTypeName: "Search Engines", type: "selectRelated" },
      { count: 3, referrerTypeName: "whatsoever", type: "visit_content" },
      { count: 1, referrerTypeName: "whatsoever", type: "whoever" },
    ];
    const countrelated = Metrics.getSelectRelated(sessionCountArray);

    expect(countrelated).toStrictEqual({
      selectRelatedCount: 8,
      visitorSelectedRelated: true,
      visitorWasRedirected: true,
    });
  });
  it("it should get user session Type", () => {
    const sessionCountArray = [
      { count: 8, referrerTypeName: "Search Engines", type: "selectRelated" },
      { count: 3, referrerTypeName: "whatsoever", type: "visit_content" },
      { count: 1, referrerTypeName: "whatsoever", type: "whoever" },
    ];
    const countrelated = Metrics.getUserType(sessionCountArray);
    expect(countrelated).toBe(true);
    console.log(countrelated);
  });
  it("it should get user session Type 2", () => {
    const sessionCountArray = [
      { count: 2, referrerTypeName: "whatsoever", type: "visit_content" },
      { count: 1, referrerTypeName: "whatsoever", type: "whoever" },
    ];
    const countrelated = Metrics.getUserType(sessionCountArray);
    expect(countrelated).toBe(false);
  });
});
