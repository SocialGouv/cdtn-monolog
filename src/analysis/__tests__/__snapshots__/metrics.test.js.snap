// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cleaning pipe shodl clean the user logs correctly 1`] = `
Array [
  Array [
    Object {
      "count": 3,
      "referrerTypeName": "Search Engines",
      "type": "visit_content",
    },
    Object {
      "count": 1,
      "referrerTypeName": "Search Engines",
      "type": "outils",
    },
  ],
  Array [
    Object {
      "count": 1,
      "referrerTypeName": "Search Engines",
      "type": "visit_content",
    },
  ],
]
`;
